src/connectMenu.js

```diff
@@ -0,0 +1,56 @@
+import React from 'react';
+
+import ContextMenuTrigger from './ContextMenuTrigger';
+import listener from './globalEventListener';
+
+// collect ContextMenuTrigger's expected props to NOT pass them on as part of the context
+const ignoredTriggerProps = [...Object.keys(ContextMenuTrigger.propTypes), 'children'];
+
+// expect the id of the menu to be responsible for as outer parameter
+export default function (menuId) {
+    // expect menu component to connect as inner parameter
+    // <Child/> is presumably a wrapper of <ContextMenu/>
+    return function (Child) {
+        // return wrapper for <Child/> that forwards the ContextMenuTrigger's additional props
+        return class ConnectMenu extends React.Component {
+            constructor(props) {
+                super(props);
+                this.state = { trigger: null };
+            }
+
+            componentDidMount() {
+                this.listenId = listener.register(this.handleShow, this.handleHide);
+            }
+
+            componentWillUnmount() {
+                if (this.listenId) {
+                    listener.unregister(this.listenId);
+                }
+            }
+
+            handleShow = (e) => {
+                if (e.detail.id !== menuId) return;
+
+                // the onShow event's detail.data object holds all ContextMenuTrigger props
+                const { data } = e.detail;
+                const filteredData = {};
+
+                for (const key in data) {
+                    // exclude props the ContextMenuTrigger is expecting itself
+                    if (!ignoredTriggerProps.includes(key)) {
+                        filteredData[key] = data[key];
+                    }
+                }
+                this.setState({ trigger: filteredData });
+            }
+
+            handleHide = () => {
+                this.setState({ trigger: null });
+            }
+
+            render() {
+                return <Child {...this.props} trigger={this.state.trigger} />;
```

Comment:
Yeah, I tried this but the structure gets all wrong.
The custom menu component (e.g. `DynamicMenu`) would return only a list of `MenuItem` on `render()` then, but react prefers only a single root element (i.e. `ContextMenu`).
I agree it would be better to avoid duplicating the menu's `id` both on the `connectMenu()` call and on the wrapped `DynamicMenu`'s root `ContextMenu`.

If there's a nice way, I cannot see it yet.